### PR TITLE
Lower GLIBC requirements by using building linux natives on ubuntu 18.04 docker

### PR DESCRIPTION
**Pull request by theofficialgman**
alternative https://github.com/libgdx/Jamepad/pull/22
closes https://github.com/libgdx/Jamepad/issues/30

### CODE REVIEW

**Comment by streamingdv** [comment]
@theofficialgman hi, just curious does the github action build still work on your side. I had recently problems on my end so I checked out your changes but still it doesn't seem to succeed building the Linux binaries. Here is a test fork of my project

https://github.com/streamingdv/Jamepad/

Are you encountering the same issues in your fork branches?
It seems the docker branch fixes the strange issue "Unable to locate package linux-libc-dev:i386" which seems to be occurring recently but fails now at _run make -j_ make not found. E.g.

https://github.com/streamingdv/Jamepad/actions/runs/6625468273/job/17996650912

**Comment by streamingdv** [comment]
@theofficialgman okay fixed it, I think something has changed in the docker build image

https://github.com/streamingdv/Jamepad/commit/c077df9371fbd876e2e1db47b1e8d8db07013c99

Fails now on the Gradle task...

**Comment by theofficialgman** [comment]
> @theofficialgman okay fixed it, I think something has changed in the docker build image
> 
> [streamingdv@c077df9](https://github.com/streamingdv/Jamepad/commit/c077df9371fbd876e2e1db47b1e8d8db07013c99)
> 
> Fails now on the Gradle task...

```
link:
     [exec] g++: error: /__/__w/Jamepad/Jamepad/SDL/build-linux32/build/.libs/libSDL2.a: No such file or directory
```
yeah strange. this is happening in jamepad master branch as well. I can reproduce on my x86_64 computer locally as well. there is no `libSDL2.a` produced by the make command.

this is strange as nothing has changed at all that I can see.

maybe gradle was previously able to use these files instead
```
│   ├── libSDL2.la
│   ├── libSDL2main.la
│   ├── libSDL2_test.la
```

**Comment by streamingdv** [comment]
Yeah I saw that, It's super strange, the make command passes now but the necessary output files weren't build. Nothing changed and I'm not sure what is causing that problem all of a sudden.

I did not yet check on a real machine, does these files maybe get outputted to a different location than previously expected?

**Comment by streamingdv** [comment]
@theofficialgman for testing purposes I just reverted the build action file to the original and disabled all Linux builds in it (only Windows x64/x86 enabled) and it seems that is working. (I removed the linux-libc-dev:i386 dependency as well)

https://github.com/streamingdv/Jamepad/blob/master/.github/workflows/pushaction.yml

So something different when make is building the Linux stuff. Not sure if it only affects Linux x86 maybe? Windows binaries can be successfully produced

https://github.com/streamingdv/Jamepad/actions/runs/6629530789

File: .github/workflows/pushaction.yml [hunk 1]
```diff
@@ -2,9 +2,6 @@ name: Compile and deploy snapshot artifacts
 
 on:
   push:
-    branches:
```
Comment (MrStahlfelge) [inline .github/workflows/pushaction.yml:5]: I still think it is good to have a restriction on master and branches named sdl/, so please keep this

**Review by MrStahlfelge — CHANGES_REQUESTED** [review]

File: .github/workflows/pushaction.yml (hunk 1 above)
Comment (theofficialgman) [reply .github/workflows/pushaction.yml:5]: I will force push it out. I need this removed to test the actions as my branch is named `docker`

**Comment by theofficialgman** [comment]
ok too me many hours but I finally realized what the actual issue is @streamingdv 
```
[exec] g++: error: /__/__w/Jamepad/Jamepad/SDL/build-linux64/build/.libs/libSDL2.a: No such file or directory
```
the issue is the `/__` the beginning of the filepath.... it was never to do with the SDL2 build itself. the issue lies with gradle

**Comment by theofficialgman** [comment]
@MrStahlfelge I have spent most of my day trying to bug test your CIs issue with the incorrect filepath issue.
This has no related to this PR as it happens on master branch now.

This PR is good for merging, I will leave it up to you to fix your CI

**Comment by MrStahlfelge** [comment]
Probably the fix is already suggested in #21

**Comment by theofficialgman** [comment]
> Probably the fix is already suggested in #21

Nope. Read again. We have no issues installing dependencies.

Issue is during linking https://github.com/libgdx/Jamepad/pull/23#issuecomment-1783585903

As I said. Fix your scripts as is (they don't work) and then you can come back to merging PRs.

**Comment by MrStahlfelge** [comment]
Build is fixed now

**Comment by theofficialgman** [comment]
@MrStahlfelge build failing in same place
https://github.com/theofficialgman/Jamepad/actions/runs/6936137715/job/18867810888#step:39:71

**Comment by streamingdv** [comment]
@theofficialgman I'm still not 100% sure what is causing the issue but I had this problem as well in my fork when I used your action script to build inside a docker container. If you are building the lib normally (without using a container) it will work and you will not run into this problem.

File: .github/workflows/pushaction.yml [hunk 2]
```diff
@@ -67,10 +94,13 @@ jobs:
     - name: Setup Gradle
       uses: gradle/gradle-build-action@v2.4.2
 
-    - run: echo "http://ports.ubuntu.com/ubuntu-ports" | sudo tee -a /etc/apt/apt-mirrors.txt
+    - run: sudo sed -i 's/deb http/deb [arch=amd64,i386] http/' /etc/apt/sources.list
+    - run: sudo grep "ubuntu.com/ubuntu" /etc/apt/sources.list | sudo tee /etc/apt/sources.list.d/ports.list
```
[the remark this answers is not in the document]
Comment (theofficialgman) [reply .github/workflows/pushaction.yml:98]: no, not sure why I put it there.

**Comment by theofficialgman** [comment]
still same linker error https://github.com/theofficialgman/Jamepad/actions/runs/8681588836/job/23804515917#step:39:74

File: .github/workflows/pushaction.yml [hunk 3]
```diff
@@ -67,10 +94,13 @@ jobs:
     - name: Setup Gradle
       uses: gradle/gradle-build-action@v2.4.2
 
-    - run: echo "http://ports.ubuntu.com/ubuntu-ports" | sudo tee -a /etc/apt/apt-mirrors.txt
+    - run: sudo sed -i 's/deb http/deb [arch=amd64,i386] http/' /etc/apt/sources.list
+    - run: grep "ubuntu.com/ubuntu" /etc/apt/sources.list | sudo tee /etc/apt/sources.list.d/ports.list
+    - run: sudo sed -i 's/amd64,i386/armhf,arm64/' /etc/apt/sources.list.d/ports.list
+    - run: sudo sed -i 's#http://.*/ubuntu#http://ports.ubuntu.com/ubuntu-ports#' /etc/apt/sources.list.d/ports.list
```
[the remark this answers is not in the document]
Comment (theofficialgman) [reply .github/workflows/pushaction.yml:100]: the default x86_64 ubuntu archive only supports amd64 and i386. It does not have armhf/arm64 and other architectures. So the 1st line limits the existing sources.list to amd64 and i386. The 2nd line copies the existing sources.list to a new file for supporting the other architectures. the 3rd line edits the architectures in the new file to armhf and arm64. and the 4th line changes the ubuntu source urls from the default urls (which a reminder only have amd64 and i386) to the ports repo which is for all the non-amd64/i386 architectures (like armhf/arm64)

two urls that display this clearly
the default ubuntu archive: http://archive.ubuntu.com/ubuntu/dists/jammy/
the ports ubuntu archive: http://ports.ubuntu.com/dists/jammy/

File: .github/workflows/pushaction.yml [hunk 4]
```diff
@@ -50,11 +50,38 @@ jobs:
...
+        openjdk-11-jdk-headless \
+        maven \
+        build-essential \
+        ant sudo locales
+        # set Locale to en_US.UTF-8 (avoids hang during compilation)
+        locale-gen en_US.UTF-8
+        echo "LANG=en_US.UTF-8" >> $GITHUB_ENV
+        echo "LANGUAGE=en_US.UTF-8" >> $GITHUB_ENV
+        echo "LC_ALL=en_US.UTF-8" >> $GITHUB_ENV    
+    - uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (theofficialgman) [reply .github/workflows/pushaction.yml:84]: Yes, unfortunately nodejs 20 doesn't support bionic (glibc 2.17) so checkout v3 was necessary.

Unfortunately supporting these older systems gets more difficult by the day.

Personally I have given up on that for my own projects but I leave PRs like this open incase upstream repositories want to continue playing that game.

**Comment by MrStahlfelge** [comment]
Now we need someone to fix the failure :)

**Comment by theofficialgman** [comment]
just an fyi this will probably fail because of the nodejs 20 forcing that is effect. you have to provide an environment variable to continue using nodejs 16

**Comment by theofficialgman** [comment]
yes because github transitioned to forcing node20 for all actions for anything that previously used node16

**Comment by theofficialgman** [comment]
> It says this though.
> 
> > This will only work until we upgrade the runner removing Node16 later in the spring.

Its still not been removed at this point and I do not thing it will ever be removed (for existing images).

Anyway I said this for a reason https://github.com/libgdx/Jamepad/pull/23#discussion_r1799703640

**Comment by theofficialgman** [comment]
I've been using it in other projects for daily CI and yet to have an instance where node16 wasn't available. Same goes for other projects. I think that node12 is even still available with the same flag (if an actions requests it).

https://github.com/Pi-Apps-Coders/box64-debs/blob/49b7664bdac080f646edff932174d412c87c0b61/.github/workflows/Update-Box64.yml#L10-L12

**Comment by theofficialgman** [comment]
The one I sent is all you need.

**Comment by MrStahlfelge** [comment]
Does the approach by setting the env variable in the build script work? I would prefer this over configuring it on the repo for documentation reasons.

**Comment by theofficialgman** [comment]
It needs to be in the yml since it's the actions commandd that need it, not any of the bash scripts.

**Comment by theofficialgman** [comment]
yeah it is. You can run containers per step but it gets to be more of a hassle than just running the entire job in a container (like this PR did) because the container isn't "saved" after a step completes

https://docs.github.com/en/actions/writing-workflows/choosing-where-your-workflow-runs/running-jobs-in-a-container

there are other people asking for functionality that would allow for runinng a sequence of steps to occur in a container while allowing some steps before/after to occur on the host but github has never implemented it https://github.com/actions/runner/issues/320#issuecomment-582854054

**Comment by MrStahlfelge** [comment]
Looks good so far, but push does not work any more which is not related to this problem

**Comment by MrStahlfelge** [comment]
@Tom-Ski is that something you can handle?

**Comment by Tom-Ski** [comment]
Should be good

**Pull request merged**